### ISSUE TITLE
Publishing messages infinite fail loop

### ISSUE THREAD

**Issue by jagofett** [dotnetcore/CAP#533]
H!

I have a problem with CAP 3.02 and publishing messages after a connection lost to RabbitMQ server.
We use CAP with postgre and RabbitMQ, and since we use microservice architecture, it can happen, that for a short period, the RabbitMQ server is not available. In that case, in our application services is often happens, that after a reconnection to the RabbitMQ, pushing message will fall, only a service restart resolve the issue.

**Steps to reproduce the issue:**

1. Microservices (publusher, receiver, RabbitMQ) is running, and connected to each other.
2. In that case, publishing works as expected (first _RabbitMQ topic message [Offer.CartChanged] has been published._ and _Received message. id:1244225502148739072, name: Offer.CartChanged_ messages
3. Stop the RabbitMQ service long before send the next message, the first 3 retry will fail, start the RabbitMQ. After the 5 min waiting, the 4th retry will be successfull - this is the expected result. In the exception you can see, that the RabbitMQ publisher return a status:320, connection closed)

4. If you stop the RabbitMQ just before the message will be sent, the firt 3 retry will be unsuccessful, but after the 5 min waiting time, the other retries neither will be succeed - and no new message can be sent, because the exception will be always thrown. (RabbitMQ publusher return status:200, Connection close forced, and never retry to connect again)
5. Restart the publisher ms, and the message will be sent.

I've attached the log files with the publisher and receiver service as well.
[sender.log](https://github.com/dotnetcore/CAP/files/4398790/sender.log)
[receiver.log](https://github.com/dotnetcore/CAP/files/4398791/receiver.log)

Db logs:
![published messages](https://user-images.githubusercontent.com/9073384/77848496-75f97480-71c5-11ea-91d8-96a3379c77d6.png)

Can you please investigate this issue? Ask me if anything is missing out.

Thanks in advance for taking time for this issue.

**Comment by jagofett** [comment]
I think I've found the problematic part.
in the **src\DotNetCore.CAP.RabbitMQ\IConnectionChannelPool.Default.cs** file, in the **Return** method:
When a connection is closed, it shouldn't placed back to the pool - it cannot be used, and it will always throw an exception, and additionally, it will never try to make a new connection. Possible solution is checking also if a connection is open, before enqueue-ing it back to the pool.
Code snippet:

```
public virtual bool Return(IModel connection)
        {
            if (Interlocked.Increment(ref _count) <= _maxSize && connection.IsOpen) //possible fix?
            {
                _pool.Enqueue(connection);

                return true;
            }

            Interlocked.Decrement(ref _count);

            Debug.Assert(_maxSize == 0 || _pool.Count <= _maxSize);

            return false;
        }
```

Can you please take a look at it?

**Comment by yang-xiaodong** [comment]
Thanks for your feedback, I will check this issue later

**Comment by yang-xiaodong** [comment]
Hello @jagofett ,
The fix you pointed looks good, simple and without side effects. let's do it

**Comment by yang-xiaodong** [comment]
Fixed in v3.0.3

**Comment by smedjes** [comment]
Hello @yang-xiaodong!
Can this be an issue for NATS as well?
I'm investigating an issue we're having that seems to be occurring after I stop and restart nats-server.

2024-06-11 08:01:04.934 +02:00 [DBG] Transport connection healthy!
2024-06-11 08:01:05.339 +02:00 [ERR] NATS server connection error. -->  NATS.Client.NATSConnectionException: Server closed the connection.
   at NATS.Client.Connection.readLoop()
2024-06-11 08:01:05.340 +02:00 [ERR] NATS server connection error. -->  NATS.Client.NATSConnectionException: Server closed the connection.
   at NATS.Client.Connection.readLoop()
2024-06-11 08:01:05.341 +02:00 [ERR] NATS server connection error. -->  NATS.Client.NATSConnectionException: Server closed the connection.
   at NATS.Client.Connection.readLoop()
2024-06-11 08:01:34.955 +02:00 [DBG] Transport connection checking...
2024-06-11 08:01:34.955 +02:00 [WRN] Transport connection is unhealthy, reconnection...



2024-06-11 08:02:04.938 +02:00 [INF] Received package.transfer.completed
2024-06-11 08:02:05.484 +02:00 [ERR] An exception occurred while publishing a message, reason:Failed : . message id:10765409202733075
DotNetCore.CAP.Internal.PublisherSentFailedException: Connection is closed.
 ---> NATS.Client.NATSConnectionClosedException: Connection is closed.
   at NATS.Client.Connection.PublishImpl(String subject, String reply, MsgHeader inHeaders, Byte[] data, Int32 offset, Nullable`1 inCount, Boolean flushBuffer)
   at NATS.Client.Connection.RequestAsyncImpl(String subject, MsgHeader headers, Byte[] data, Int32 offset, Nullable`1 count, Int32 timeout, CancellationToken token)
   at NATS.Client.JetStream.JetStream.PublishAsyncInternal(String subject, Byte[] data, MsgHeader hdr, PublishOptions options)
   at DotNetCore.CAP.NATS.NATSTransport.SendAsync(TransportMessage message)


So I'm able to receive messages, but when I try to publish I get a error saying the connection is closed.
Do you have any ideas?

Thanks in advance!
